### ISSUE TITLE
Crash with `algorithmic_multiplication_binary40` on `neural_gpu`

### ISSUE THREAD

**Issue by schani** [tensorflow/tensor2tensor#80]
I'm using tensor2tensor 1.0.9, running this script:

```
PROBLEM=algorithmic_multiplication_binary40
MODEL=neural_gpu
HPARAMS=neuralgpu_1

DATA_DIR=/data/t2t-test/t2t_data
TMP_DIR=/tmp/t2t_datagen
TRAIN_DIR=/data/t2t-test/t2t_train/$PROBLEM/$MODEL-$HPARAMS

# Train                                                                                                                                                                                                                         
# *  If you run out of memory, add --hparams='batch_size=2048' or even 1024.                                                                                                                                                    
t2t-trainer \
    --data_dir=$DATA_DIR \
    --problems=$PROBLEM \
    --model=$MODEL \
    --hparams_set=$HPARAMS \
    --output_dir=$TRAIN_DIR
```

After training for a while I get this:

```
INFO:tensorflow:step = 1501, loss = 0.826424 (74.946 sec)
INFO:tensorflow:Saving checkpoints for 1588 into /data/t2t-test/t2t_train/algorithmic_multiplication_binary40/neural_gpu-neuralgpu_1/model.ckpt.
INFO:tensorflow:global_step/sec: 1.30914
INFO:tensorflow:step = 1601, loss = 0.803077 (76.385 sec)
INFO:tensorflow:global_step/sec: 1.29677
INFO:tensorflow:step = 1701, loss = 0.796181 (77.115 sec)
INFO:tensorflow:global_step/sec: 1.28511
INFO:tensorflow:step = 1801, loss = 0.76857 (77.814 sec)
INFO:tensorflow:global_step/sec: 1.28044
INFO:tensorflow:step = 1901, loss = 0.743739 (78.098 sec)
INFO:tensorflow:datashard_devices: ['gpu:0']
INFO:tensorflow:caching_devices: None
Traceback (most recent call last):
  File "/home/ubuntu/tensorflow/bin/t2t-trainer", line 83, in <module>
    tf.app.run()
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "/home/ubuntu/tensorflow/bin/t2t-trainer", line 79, in main
    schedule=FLAGS.schedule)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/trainer_utils.py", line 246, in run
    run_locally(exp_fn(output_dir))
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/trainer_utils.py", line 536, in run_locally
    exp.train_and_evaluate()
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 495, in train_and_evaluate
    self.train(delay_secs=0)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 275, in train
    hooks=self._train_monitors + extra_hooks)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 665, in _call_train
    monitors=hooks)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/util/deprecation.py", line 289, in new_func
    return func(*args, **kwargs)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 455, in fit
    loss = self._train_model(input_fn=input_fn, hooks=hooks)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1007, in _train_model
    _, loss = mon_sess.run([model_fn_ops.train_op, model_fn_ops.loss])
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 505, in run
    run_metadata=run_metadata)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 842, in run
    run_metadata=run_metadata)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 798, in run
    return self._sess.run(*args, **kwargs)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/training/monitored_session.py", line 960, in run
    run_metadata=run_metadata))
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/monitors.py", line 1178, in after_run
    induce_stop = m.step_end(self._last_step, result)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/monitors.py", line 357, in step_end
    return self.every_n_step_end(step, output)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/monitors.py", line 677, in every_n_step_end
    validation_outputs = self._evaluate_estimator()
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/monitors.py", line 653, in _evaluate_estimator
    metrics=self.metrics, hooks=self.hooks, name=self.name)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/util/deprecation.py", line 289, in new_func
    return func(*args, **kwargs)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 543, in evaluate
    log_progress=log_progress)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 829, in _evaluate_model
    model_fn_results = self._get_eval_ops(features, labels, metrics)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1190, in _get_eval_ops
    features, labels, model_fn_lib.ModeKeys.EVAL)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1133, in _call_model_fn
    model_fn_results = self._model_fn(features, labels, **kwargs)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/trainer_utils.py", line 431, in model_fn
    len(hparams.problems) - 1)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/trainer_utils.py", line 750, in _cond_on_index
    return fn(cur_idx)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/trainer_utils.py", line 412, in nth_model
    features, skip=(skipping_is_on and skip_this_one))
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/t2t_model.py", line 415, in model_fn
    transformed_features)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/t2t_model.py", line 464, in model_fn_body_sharded
    datashard_to_features)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/expert_utils.py", line 294, in __call__
    outputs.append(fns[i](*my_args[i], **my_kwargs[i]))
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/utils/t2t_model.py", line 40, in fn_with_timing
    res = fn(*args, **kwargs)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/models/neural_gpu.py", line 60, in model_fn_body
    return neural_gpu(features["inputs"], self._hparams)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensor2tensor/models/neural_gpu.py", line 53, in neural_gpu
    swap_memory=True)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/ops/functional_ops.py", line 122, in foldl
    swap_memory=swap_memory)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2770, in while_loop
    result = context.BuildLoop(cond, body, loop_vars, shape_invariants)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2599, in BuildLoop
    pred, body, original_loop_vars, loop_vars, shape_invariants)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2580, in _BuildLoop
    _EnforceShapeInvariant(m_var, n_var)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/ops/control_flow_ops.py", line 575, in _EnforceShapeInvariant
    % (merge_var.name, m_shape, n_shape))
ValueError: The shape for body/model/parallel_0/body/neural_gpu/foldl/while/Merge_1:0 is not an invariant for the loop. It enters the loop with shape (?, ?, 1, 256), but has shape (?, ?, ?, 256) after one iteration. Provide shape invariants using either the `shape_invariants` argument of tf.while_loop or set_shape() on the loop variables.
```

**Comment by kolloldas** [comment]
Yes I have found this bug too. The _conv_internal()_ function in _common_layers.py_ is explicitly modifying the shape:

```
  # Manually setting the shape to be unknown in the middle two dimensions so
  # that the `tf.cond` below won't throw an error based on the convolution
  # kernels being too large for the data.
  inputs._shape = tf.TensorShape([static_shape[0], None, None, static_shape[3]])  # pylint: disable=protected-access
```
If after the call I restore the original shape the bug disappears:
```
def conv_internal(conv_fn, inputs, filters, kernel_size, **kwargs):
  """Conditional conv_fn making kernel 1d or 2d depending on inputs shape."""
  static_shape = inputs.get_shape()
  
  ...
  
  # Manually setting the shape to be unknown in the middle two dimensions so
  # that the `tf.cond` below won't throw an error based on the convolution
  # kernels being too large for the data.
  inputs._shape = tf.TensorShape([static_shape[0], None, None, static_shape[3]])  # pylint: disable=protected-access
  if kernel_size[1] == 1 or force2d:
    # Avoiding the cond below can speed up graph and gradient construction.
    c = conv2d_kernel(kernel_size, "single")
  else: c = tf.cond(
      tf.equal(tf.shape(inputs)[2],
               1), lambda: conv2d_kernel((kernel_size[0], 1), "small"),
      lambda: conv2d_kernel(kernel_size, "std"))
  inputs.set_shape(static_shape)
  c.set_shape(static_shape)
  return c
```
Is there a better way to fix this?

**Comment by lukaszkaiser** [comment]
This is a good fix. I believe it might be even better to be more radical and remove the tf.cond altogether -- we used it in some early experiments, but it never worked well. It's probably better to just let it go, I don't think any current models use it.

**Comment by kolloldas** [comment]
Removing the cond altogether would be better since the fix I gave above assumes the input and output shapes remain same which need not be true. In fact several unit tests fail with the fix. 

However upon removing the cond (and my fix) two tests are still failing (there is a 3rd unrelated one):

```
======================================================================
ERROR: testConvGRU (__main__.CommonLayersTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tensor2tensor/models/common_layers_test.py", line 229, in testConvGRU
    tf.constant(x, dtype=tf.float32), (1, 3), 11, padding="LEFT")
...
 File "/mnt/e/dev/venv/t2t/lib/python3.4/site-packages/tensorflow/python/layers/convolutional.py", line 127, in build
    raise ValueError('The channel dimension of the inputs '
ValueError: The channel dimension of the inputs should be defined. Found `None`.
```

and

```
======================================================================
ERROR: testConvStride2MultiStep (__main__.CommonLayersTest)
----------------------------------------------------------------------
...
Traceback (most recent call last):
  File "tensor2tensor/models/common_layers_test.py", line 283, in testConvStride2MultiStep
    tf.constant(x1, dtype=tf.float32), 4, 16)
...
  File "/mnt/e/dev/venv/t2t/lib/python3.4/site-packages/tensorflow/python/framework/common_shapes.py", line 676, in _call_cpp_shape_fn_impl
    raise ValueError(err.message)
ValueError: Negative dimension size caused by subtracting 2 from 1 for 'conv_stride2_multistep/conv0_single/convolution' (op: 'Conv2D') with input shapes: [5,32,1,11], [2,2,11,16].

```

The first fail can be fixed by setting padding to VALID/SAME. Couldn't figure out the 2nd (the test expects a 0 in the 3rd dimension). Perhaps the test cases need to be updated themselves.

**Comment by lukaszkaiser** [comment]
I think it's ok to disable the second test for now. Will you send a PR? Thanks for doing this!

**Comment by schani** [comment]
Unfortunately, applying `set_shape` in either case makes transformer fail for me on `mnist_test`:

```
Traceback (most recent call last):
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/framework/tensor_shape.py", line 558, in merge_with
    new_dims.append(dim.merge_with(other[i]))   
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/framework/tensor_shape.py", line 133, in merge_with
    self.assert_is_compatible_with(other)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/framework/tensor_shape.py", line 106, in assert_is_compatible_with
    other))
ValueError: Dimensions 1536 and 512 are not compatible

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/tensorflow/bin/t2t-trainer", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/ubuntu/tensor2tensor/tensor2tensor/bin/t2t-trainer", line 83, in <module>
    tf.app.run()
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "/home/ubuntu/tensor2tensor/tensor2tensor/bin/t2t-trainer", line 79, in main
    schedule=FLAGS.schedule)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/trainer_utils.py", line 246, in run
    run_locally(exp_fn(output_dir))
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/trainer_utils.py", line 536, in run_locally
    exp.train_and_evaluate()
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 495, in train_and_evaluate
    self.train(delay_secs=0)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 275, in train
    hooks=self._train_monitors + extra_hooks)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 665, in _call_train
    monitors=hooks)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/util/deprecation.py", line 289, in new_func
    return func(*args, **kwargs)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 455, in fit
    loss = self._train_model(input_fn=input_fn, hooks=hooks)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 955, in _train_model
    model_fn_ops = self._get_train_ops(features, labels)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1162, in _get_train_ops
    return self._call_model_fn(features, labels, model_fn_lib.ModeKeys.TRAIN)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1133, in _call_model_fn
    model_fn_results = self._model_fn(features, labels, **kwargs)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/trainer_utils.py", line 431, in model_fn
    len(hparams.problems) - 1)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/trainer_utils.py", line 750, in _cond_on_index
    return fn(cur_idx)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/trainer_utils.py", line 412, in nth_model
    features, skip=(skipping_is_on and skip_this_one))
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/t2t_model.py", line 415, in model_fn
    transformed_features)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/t2t_model.py", line 464, in model_fn_body_sharded
    datashard_to_features)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/expert_utils.py", line 294, in __call__
    outputs.append(fns[i](*my_args[i], **my_kwargs[i]))
  File "/home/ubuntu/tensor2tensor/tensor2tensor/utils/t2t_model.py", line 40, in fn_with_timing
    res = fn(*args, **kwargs)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/models/transformer.py", line 68, in model_fn_body
    encoder_attention_bias, hparams)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/models/transformer.py", line 165, in transformer_encoder
    name="encoder_self_attention"))
  File "/home/ubuntu/tensor2tensor/tensor2tensor/models/common_attention.py", line 388, in multihead_attention
    name="qkv_transform")
  File "/home/ubuntu/tensor2tensor/tensor2tensor/models/common_layers.py", line 351, in conv1d
    conv(tf.expand_dims(inputs, 2), filters, (kernel_size, 1), **kwargs), 2)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/models/common_layers.py", line 346, in conv
    return conv_internal(tf.layers.conv2d, inputs, filters, kernel_size, **kwargs)
  File "/home/ubuntu/tensor2tensor/tensor2tensor/models/common_layers.py", line 341, in conv_internal
    c.set_shape(static_shape)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 443, in set_shape
    self._shape = self._shape.merge_with(shape)
  File "/home/ubuntu/tensorflow/lib/python3.5/site-packages/tensorflow/python/framework/tensor_shape.py", line 561, in merge_with
    raise ValueError("Shapes %s and %s are not compatible" % (self, other))
ValueError: Shapes (?, ?, ?, 1536) and (?, ?, 1, 512) are not compatible
```

If I don't apply @kolloldas change, the model is created successfully.  `mnist` still doesn't run, but for other reasons.

**Comment by kolloldas** [comment]
@schani I agree the set_shape() call is not the right fix. Also removing just the cond in _conv_internal()_ won't do. There is a cond inside the _pool()_ function as well which needs to be removed.

@lukaszkaiser is it ok to replace this cond as well?

```
def pool(inputs, window_size, pooling_type, padding, strides=(1, 1)):
...
    # Manually setting the shape to be unknown in the middle two dimensions so
    # that the `tf.cond` below won't throw an error based on the convolution
    # kernels being too large for the data.
    inputs._shape = tf.TensorShape(  # pylint: disable=protected-access
        [static_shape[0], None, None, static_shape[3]])
    return tf.cond(
        tf.equal(tf.shape(inputs)[2], 1),
        lambda: tf.nn.pool(  # pylint: disable=g-long-lambda
            inputs, window_size_small, pooling_type, padding,
            strides=strides_small),
        lambda: tf.nn.pool(  # pylint: disable=g-long-lambda
            inputs, window_size, pooling_type, padding, strides=strides))
```
with:

```
def pool(inputs, window_size, pooling_type, padding, strides=(1, 1)):
  """Pooling (supports "LEFT")."""
  ...

    return tf.nn.pool( inputs, window_size, pooling_type, padding, strides=strides)
```

If I remove both then the model is built although transformer still fails on the image_mnist_test task ( It doesn't currently work on 2D data I believe)

If these changes are fine I'll send a PR..

**Comment by lukaszkaiser** [comment]
Yes, let's get rid of both conds. Transformer wasn't built for MNIST, but I'll get to making it work there too. Step by step, let's start with removing the conds.

**Comment by kolloldas** [comment]
#92

**Comment by lukaszkaiser** [comment]
Thanks @kolloldas -- your PR was merged and is part of 1.0.11. I hope that resolves the issue, please reopen if you see this problem again!
